extensions/amp-viz-vega/0.1/amp-viz-vega.js

```diff
@@ -0,0 +1,168 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isJsonScriptTag, childElementsByTag} from '../../../src/dom';
+import {isExperimentOn} from '../../../src/experiments';
+import {tryParseJson} from '../../../src/json';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {vsyncFor} from '../../../src/vsync';
+
+/** @const */
+const EXPERIMENT = 'amp-viz-vega';
+
+export class AmpVizVega extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.container_ = null;
+
+    /** @private {?JSONType} */
+    this.data_ = null;
+
+    /** @private {?string} */
+    this.dataUrl_ = null;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * Create the vega container.
+   * Called lazily in the first `#layoutCallback`.
+   */
+  initialize_() {
+    if (this.container_) {
+      return;
+    }
+
+    this.container_ = this.element.ownerDocument.createElement('div');
+    this.dataUrl_ = this.element.getAttribute('data-url');
+
+    this.applyFillContent(this.container_, true);
+    this.element.appendChild(this.container_);
+  }
+
+  /** @override */
+  buildCallback() {
+    user.assert(isExperimentOn(this.win, EXPERIMENT),
+      `Experiment ${EXPERIMENT} disabled`);
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.initialize_();
+    return this.loadData_().then(() => this.renderGraph_());
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  loadData_() {
+    if (this.data_) {
+      return Promise.resolve();
+    }
+    const inlineData = this.getInlineData_();
+    if (!inlineData && !this.dataUrl_) {
+      const err = this.getName_() + ' neither `data-url` attribute nor a ' +
+        'valid <script type="application/json"> child was found for Vega data';
+      return Promise.reject(user.createError(err));
+    }
+
+    if (inlineData && this.dataUrl_) {
+      const err = this.getName_() + ' both `data-url` attribute and a valid ' +
+        '<script type="application/json"> child were found for Vega data. ' +
+        'Only one way of specifying the data is allowed.';
+      return Promise.reject(user.createError(err));
+    }
+
+    if (this.dataUrl_) {
+      // TODO(aghassemi): Fetch and validate the data file.
+      this.data_ = {
+        'url': this.dataUrl_,
+      };
+      return Promise.resolve();
+    }
+
+    if (inlineData) {
+      this.data_ = inlineData;
+      return Promise.resolve();
+    }
+  }
```

Comment:
I think the final return is not reachable here? If so, let's make this obvious. E.g.

```
this.data_ = dev.assert(inlineData);
return Promise.resolve();
```